[release/3.1] Redo `yarn install`

dougbu opened the pull request:
- get latest versions

dougbu committed:
[release/3.1] Redo `yarn install`
- get latest versions

dougbu commented on src/SignalR/clients/ts/signalr/package.json:
```diff
@@ -55,8 +55,5 @@
     "eventsource": "^1.0.7",
```
This is the only mention of `eventsource` in a package.json file in this branch. The mention of `@types/eventsource` above is also unique. Should one or both be removed while I'm at it❔

wtgodbe replied:
I think this is supposed to be here - should just update it to 1.1.1

dougbu replied:
I'll just leave it alone because yarn.lock shows 1.1.2 anyhow.

dougbu commented:
@Pilchie should we treat this as `tell-mode` and should I notify Tactics❔

BrennanConroy commented:
> Build break may be due to an old version of `npm` or `yarn`, I'm not sure. Suggestions❔

Looks like it's because >12.20.20 of @types/node package need Typescript version 3.7+
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/55430

dougbu commented:
> > Build break may be due to an old version of `npm` or `yarn`, I'm not sure. Suggestions❔

> Looks like it's because >12.20.20 of @types/node package need Typescript version 3.7+
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/55430

Any downaide to updating the TS version❔

BrennanConroy commented:
Probably new warnings/errors showing up in the code, but it should be fine otherwise, webpack should be controlling how the app is packaged and what format (ES5 etc.) to use.

dougbu commented:
Also hitting lots of version resolution issues w/ System.Text.Encodings.Web. Did you make a change in that area recently @wtgodbe❔ Any issues if I pin the newer version in the affected bunch o' projects❔

@dotnet/aspnet-blazor-eng the other warning about STEW is
``` text
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(2203,5): warning MSB3277: Found conflicts between different versions of "System.Text.Encodings.Web" that could not be resolved. [D:\a\_work\1\s\src\Components\Ignitor\src\Ignitor.csproj]
```
But we create a non-shipping `Ignitor` package. Can the STEW version be pinned there too❔

----

/fyi @BrennanConroy most of those issues are in SignalR test, benchmarks, and sample projects.

dougbu commented:
Going back to the `npm` / TypeScript issue, I think the new errors are related primarily to our `"request": "^2.88.0"` dependency or the `"@types/request": "^2.47.1"` dev dependency in this branch, specifically the src/SignalR/clients/ts/signalr/package.json file. We don't have either in 'main' and also don't hit these errors there. Since the `request` package has been deprecated, should I remove them from package.json instead of bumping the `typescript` version❔

BrennanConroy commented:
> Since the `request` package has been deprecated, should I remove them from package.json

No, there is a reason we depend on it

dougbu commented:
Well, it is request that's bringing in too new a version of `@types/node` for TypeScript:

``` yaml
"@types/request@^2.47.1":
  version "2.48.8"
  resolved "https://registry.yarnpkg.com/@types/request/-/request-2.48.8.tgz#0b90fde3b655ab50976cb8c5ac00faca22f5a82c"
  integrity sha512-whjk1EDJPcAR2kYHRbFl/lKeeKYTi05A15K9bnLInCVroNDCtXce57xKdI0/rQaA3K+6q0eFyUBPmqfSndUZdQ==
  dependencies:
    "@types/caseless" "*"
    "@types/node" "*"
    "@types/tough-cookie" "*"
    form-data "^2.5.0"
```

I'll bump `typescript` since it's a dev dependency and hope for the best

dougbu committed:
[release/3.1] Use latest STEW everywhere
- avoid package resolution (`MSB3277`) warnings

Pilchie commented:
I think this and related changes are fine for `tell-mode`

BrennanConroy commented:
> src/HandshakeProtocol.ts(42,59): error TS2345: Argument of type 'Uint8Array' is not assignable to parameter of type 'number[]'. [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/HandshakeProtocol.ts(42,59): error TS2345: Argument of type 'Uint8Array' is not assignable to parameter of type 'number[]'.
    Type 'Uint8Array' is missing the following properties from type 'number[]': pop, push, concat, shift, and 3 more.
src/HttpConnection.ts(437,29): error TS2322: Type 'unknown' is not assignable to type 'Error | ITransport'. [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/HttpConnection.ts(437,29): error TS2322: Type 'unknown' is not assignable to type 'Error | ITransport'.
src/HubConnection.ts(232,40): error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Error | undefined'. [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/HubConnection.ts(232,40): error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Error | undefined'.
src/HubConnection.ts(740,13): error TS2774: This condition will always return true since this function is always defined. Did you mean to call it instead? [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/HubConnection.ts(740,13): error TS2774: This condition will always return true since this function is always defined. Did you mean to call it instead?
src/HubConnection.ts(773,21): error TS2774: This condition will always return true since this function is always defined. Did you mean to call it instead? [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/HubConnection.ts(773,21): error TS2774: This condition will always return true since this function is always defined. Did you mean to call it instead?
src/HubConnection.ts(794,65): error TS2571: Object is of type 'unknown'. [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/HubConnection.ts(794,65): error TS2571: Object is of type 'unknown'.
src/LongPollingTransport.ts(153,113): error TS2571: Object is of type 'unknown'. [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/LongPollingTransport.ts(153,113): error TS2571: Object is of type 'unknown'.
src/LongPollingTransport.ts(160,29): error TS2322: Type 'unknown' is not assignable to type 'Error | undefined'. [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/LongPollingTransport.ts(160,29): error TS2322: Type 'unknown' is not assignable to type 'Error | undefined'.
src/ServerSentEventsTransport.ts(75,40): error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Error | undefined'. [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/ServerSentEventsTransport.ts(75,40): error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Error | undefined'.
src/ServerSentEventsTransport.ts(81,17): error TS2322: Type '(e: MessageEvent) => void' is not assignable to type '(this: EventSource, ev: Event) => any'. [D:\a\_work\1\s\src\SignalR\clients\ts\signalr\signalr.npmproj]
##[error]src/ServerSentEventsTransport.ts(81,17): error TS2322: Type '(e: MessageEvent) => void' is not assignable to type '(this: EventSource, ev: Event) => any'.
    Types of parameters 'e' and 'ev' are incompatible.
      Type 'Event' is missing the following properties from type 'MessageEvent<any>': data, lastEventId, origin, ports, and 2 more.

Like I said, errors 😆

dougbu commented:
Glad you're pleased @BrennanConroy 😃 I could either mosey through the specific errors or try porting the portions of 739f58574e8 that updated the `src/SignalR/clients/ts/signalr/src/*.ts` files in 'main'. Thoughts❔

BrennanConroy commented:
That change isn't going to be helpful, Typescript was already at 4.x by then.
https://github.com/dotnet/aspnetcore/pull/29905 is the best you'd be able to draw inspiration from

dougbu commented:
> #29905 is the best you'd be able to draw inspiration from

I'm inspired but won't get back to it for a couple of hours.

@vseanreesermsft @mmitche how much time do I have to get release/3.1 building today❔

dougbu commented:
This isn't quite enough but I'll look at the remaining errors tomorrow

``` text
src/HttpConnection.ts(437,29): error TS2322: Type 'unknown' is not assignable to type 'Error | ITransport'. [...\src\SignalR\clients\ts\signalr\signalr.npmproj]
src/HubConnection.ts(232,40): error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Error |undefined'. [...\src\SignalR\clients\ts\signalr\signalr.npmproj]
src/HubConnection.ts(794,65): error TS2571: Object is of type 'unknown'. [...\src\SignalR\clients\ts\signalr\signalr.npmproj]
src/LongPollingTransport.ts(153,113): error TS2571: Object is of type 'unknown'. [...\src\SignalR\clients\ts\signalr\signalr.npmproj]
src/LongPollingTransport.ts(160,29): error TS2322: Type 'unknown' is not assignable to type 'Error | undefined'. [...\src\SignalR\clients\ts\signalr\signalr.npmproj]
src/ServerSentEventsTransport.ts(75,40): error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Error | undefined'. [...\src\SignalR\clients\ts\signalr\signalr.npmproj]
```

dougbu commented:
Current iteration will likely fail w/ `Module` versus `module` problems in src/Components/*.ts files. I think 19252d64d9cc may contain some of what's needed to fix things but that looks like a fairly major refactoring and might depend on WASM changes that weren't made here. Looking for and greatly appreciating help from @dotnet/aspnet-blazor-eng because I don't understand TypeScript versions enough to know what may have busted things.

dougbu commented:
Errors already visible in macOS build:
``` text
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Boot.WebAssembly.ts(68,16)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Boot.WebAssembly.ts(68,42)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Boot.WebAssembly.ts(70,7)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
...
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(17,44)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(18,44)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(19,46)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(24,20)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(29,46)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(67,27)
        TS2304: Cannot find name 'Module'.
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(188,54)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(243,5)
        TS2322: Type '(arg_0: number) => number' is not assignable to type '(managedString: System_String) => Pointer'.
    Types of parameters 'arg_0' and 'managedString' are incompatible.
      Type 'System_String' is not assignable to type 'number'.
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(319,27)
        TS2552: Cannot find name 'Module'. Did you mean 'module'?
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts(320,41)
        TS2304: Cannot find name 'Module'.
...
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/TimezoneDataFile.ts(32,5)
        TS2304: Cannot find name 'Module'.
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/TimezoneDataFile.ts(34,7)
        TS2304: Cannot find name 'Module'.
  [tsl] ERROR in /Users/runner/work/1/s/src/Components/Web.JS/src/Platform/Mono/TimezoneDataFile.ts(38,7)
        TS2304: Cannot find name 'Module'.
```

dougbu commented:
@BrennanConroy I also see a lot of new apparent syntax errors starting at https://dev.azure.com/dnceng/public/_build/results?buildId=1825044&view=logs&j=f43b187f-6c15-531c-a59b-e4fba4e9ed20&t=9e357cb9-1774-5509-1b39-7d8136be93e2&l=1122 with

``` text
  FAIL signalr/tests/HubConnection.test.ts
    ● Test suite failed to run
  
      TypeScript diagnostics (customize using `[jest-config].globals.ts-jest.diagnostics` option):
      signalr/tests/HubConnection.test.ts:25:1 - error TS2582: Cannot find name 'describe'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
```
Suggestions appreciated❕

BrennanConroy commented on src/SignalR/clients/ts/jest.config.js:
```diff
@@ -4,7 +4,7 @@
 module.exports = {
     globals: {
         "ts-jest": {
-            "tsConfig": "./tsconfig.jest.json",
```
Probably need to revert this and the same change in FunctionalTests, it is only relevant when updating ts-jest to v26 or v27

dougbu commented on src/SignalR/clients/ts/FunctionalTests/func.jest.config.js:
```diff
@@ -8,7 +8,7 @@ module.exports = {
     globals: {
         "ts-jest": {
-            "tsConfig": "./tsconfig.jest.json",
+            "tsconfig": "./tsconfig.jest.json",
```
```suggestion
            "tsConfig": "./tsconfig.jest.json",
```

dougbu commented on src/SignalR/clients/ts/jest.config.js:
```diff
@@ -4,7 +4,7 @@
     globals: {
         "ts-jest": {
-            "tsConfig": "./tsconfig.jest.json",
+            "tsconfig": "./tsconfig.jest.json",
```
```suggestion
            "tsConfig": "./tsconfig.jest.json",
```

japui commented:
versão/3.1] Refazer yarn install #

dougbu commented:
@dotnet/aspnet-blazor-eng please comment on the many problems compiling in src/Components/Web.JS/Microsoft.AspNetCore.Components.Web.JS.npmproj w/ the new TypeScript version. That's the remaining build problem in this PR.

dougbu commented:
@BrennanConroy some SignalR tests are also failing, specifically in client-ts.npmproj and SignalR.Npm.FunctionalTests.npmproj. I suspect the issues are compilation errors that aren't caught when building. But one message worries me:

``` text
ts-jest[versions] (WARN) Version 4.7.3 of typescript installed has not been tested with ts-jest. If you're experiencing issues, consider using a supported version (>=2.7.0 <4.0.0). Please do not report issues in ts-jest if you are using unsupported versions.
```

Do we need to update ts-jest too❔

BrennanConroy commented:
Let's step back here, what is the overall goal? Is it just to fixup things like getting eventsource to be version 1.1.2? If so we should try not to update the entire world in a patch branch and do the minimal updates needed.

dougbu committed:
[release/3.1] Limit max. `@types/node` version
- avoid version that dropped support for TypeScript < 3.7
- see DefinitelyTyped/DefinitelyTyped#55430

dougbu commented on src/SignalR/clients/ts/signalr/package.json:
```diff
@@ -57,6 +57,7 @@
   "resolutions": {
-    "url-parse": ">=1.5.8"
+    "//": "Avoid version that drops support for TypeScript < 3.7",
+    "@types/node": "<12.20.21"
```
With this and removing `"@types/node": "^10.9.4"` under `devDependencies` above, `12.20.22` gets resolved everywhere. Is that better or worse than current settings, which resolve `10.17.60` for everything except `request` (which gets `12.20.20`)❔

dougbu reviewed (commented):
Trying my alternative to bumping the `typescript` version&hellip;

dougbu committed:
!fixup! Latest yarn.lock versions

dougbu committed:
!fixup! Pin `@types/node` version

dougbu committed:
!fixup! Get Web.JS building

dougbu committed:
!fixup! Get SignalR functional tests building

dougbu commented:
General story w/ the last few commits is that we use old enough versions of `typescript` that `yarn install` tends to grab unusable versions of various packages. Seems more than just `@types/node` drop compatibility w/ `typescript` packages of a certain age in their patch or minor version updates ☹️

dougbu commented:
@vseanreesermsft @mmitche I don't remember enabling auto-merge and am sorry this went in. Did the new commit cause problems anywhere❔ Should I revert it or perhaps force `push` the branches back the way they were❔
